judge/runpipe.cc

```diff
@@ -611,6 +631,46 @@ struct state_t {
     child_exited_pipe = read_end;
   }
 
+ // Install a handler for the SIGUSR1 signal. The handler will send a byte to
+ // a pipe notifying the main loop that the child was terminated due to time limit.
+ // This method can be called only once.
+ // TODO: Refactor code to avoid code duplication with install_sigchld_handler.
+ void install_sigusr1_handler() {
+     fd_t fds[2];
+     if (pipe2(fds, O_CLOEXEC | O_NONBLOCK)) {
+         error(errno, "creating exit pipes");
+     }
+
+     // The lambda below cannot capture anything, otherwise it couldn't be made
+     // into a function pointer. Therefore the write_end must have a static
+     // lifetime.
+     fd_t read_end = fds[0];
+     static fd_t write_end = -1;
+     if (write_end != -1) {
+         error(0, "install_sigchld_handler can be called only once");
+     }
+     write_end = fds[1];
+
+     logmsg(LOG_DEBUG, "exit handler will send event using %d -> %d", write_end,
+            read_end);
+
+     signal(SIGUSR1, [](int) {
```

Comment:
I didn't dive into this in detail, but I think it would be better to use `sigaction` since it specifies behaviour more explicitly. In this case I guess it's not a big deal because `SIGUSR1` should only ever be delivered once.

Reply:
After reading https://en.cppreference.com/w/cpp/utility/program/signal I think we are fine and this code snippet is in line with the one where we handle `SIGCHLD`.